[GOBBLIN-1938] preserve x bit in manifest file based copy

arjun4084346 opened the pull request:
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!


### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1938


### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
This PR adds a post publish step in ManifestDatasetFinder which will save the source directories permissions and set them on the destination side. The current behavior always set execute bit on destination directories; the effect of this PR will be that it will unset it, if it is not set on the source side.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
updated the existing unit test

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

ZihanLi58 commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/ManifestBasedDataset.java:
```diff
@@ -128,6 +143,12 @@ public Iterator<FileSet<CopyEntity>> getFileSetIterator(FileSystem targetFs, Cop
+
+      Properties props = new Properties();
+      props.setProperty(SetPermissionCommitStep.STOP_ON_ERROR_KEY, "true");
+      CommitStep setPermissionCommitStep = new SetPermissionCommitStep(targetFs, ancestorOwnerAndPermissions, props);
```
Why do we need this step? if dir already exist on dst and user want to change the permission for that dir, it will appear in the manifest? If dir not exist, we should directly create parent dir to be the same as source dir?

arjun4084346 replied:
If dir does not exist, we create it, but we create it with execute bit set. I do not want to not do that, because execute bit may be necessary to do publish.
So, in this PR, I had to add a post publish step to set the original permission.
Honestly, the requirement for this work, unsetting execute bit, seems quite odd.

ZihanLi58 commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/publisher/CopyDataPublisher.java:
```diff
@@ -300,6 +300,10 @@ private void publishFileSet(CopyEntity.DatasetAndPartition datasetAndPartition,
 
+    // execute post publish commit steps after preserving file attributes, because some post publish step,
+    // e.g. SetPermissionCommitStep needs to set permissions
+    executeCommitSequence(postPublish);
```
We do move path from tmp to dst in this step, I believe it should happen before the set file attribute. We should make sure set file attribute preserve the right permission

arjun4084346 replied:
moving from tmp(writer.output.dir actually) to dst is publish step, right?

ZihanLi58 commented:
build test fail, also please add description to the PR to help us better understand it in the future

arjun4084346 commented:
> build test fail, also please add description to the PR to help us better understand it in the future

fixed test. updated the "draft" PR now. still not able to log into issues.apache.org and create a ticket
